com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/PollServiceInProgessOperationsExecution.java

```diff
@@ -0,0 +1,86 @@
+import org.cloudfoundry.client.lib.domain.CloudEvent;
+import org.flowable.engine.delegate.DelegateExecution;
+
+import com.sap.cloud.lm.sl.cf.client.lib.domain.CloudServiceExtended;
+import com.sap.cloud.lm.sl.cf.core.cf.clients.EventsGetter;
+import com.sap.cloud.lm.sl.cf.core.cf.clients.ServiceGetter;
+import com.sap.cloud.lm.sl.cf.core.cf.services.ServiceOperation;
+import com.sap.cloud.lm.sl.cf.core.cf.services.ServiceOperationState;
+import com.sap.cloud.lm.sl.cf.core.cf.services.ServiceOperationType;
+import com.sap.cloud.lm.sl.cf.process.util.ServiceOperationExecutor;
+
+public class PollServiceInProgessOperationsExecution extends PollServiceOperationsExecution implements AsyncExecution {
+
+    private ServiceOperationExecutor serviceOperationExecutor = new ServiceOperationExecutor();
+    private ServiceGetter serviceInstanceGetter;
+    private EventsGetter eventsGetter;
+
+    public PollServiceInProgessOperationsExecution(ServiceGetter serviceInstanceGetter, EventsGetter eventsGetter) {
+        this.serviceInstanceGetter = serviceInstanceGetter;
+        this.eventsGetter = eventsGetter;
+    }
+
+    @Override
+    protected List<CloudServiceExtended> computeServicesToPoll(ExecutionWrapper execution,
+        Map<String, ServiceOperationType> triggeredServiceOperations) {
+        List<CloudServiceExtended> existingServices = getServices(execution.getContext());
+        return getServicesWithTriggeredOperations(existingServices, triggeredServiceOperations);
+    }
+
+    private List<CloudServiceExtended> getServices(DelegateExecution context) {
+        List<CloudServiceExtended> allServicesToPoll = StepsUtil.getServicesData(context);
+        // There's no need to poll the user-provided services, because it is done synchronously:
+        return allServicesToPoll.stream()
+            .filter(service -> !service.isUserProvided())
+            .collect(Collectors.toList());
+    }
+
+    @Override
+    protected ServiceOperation getLastServiceOperation(ExecutionWrapper execution, CloudControllerClient client,
+        CloudServiceExtended service) {
+        Map<String, Object> cloudServiceInstance = serviceOperationExecutor.executeServiceOperation(service,
+            () -> serviceInstanceGetter.getServiceInstanceEntity(client, service.getName(), StepsUtil.getSpaceId(execution.getContext())),
+            execution.getStepLogger());
+
+        if (cloudServiceInstance == null || cloudServiceInstance.isEmpty()) {
+            ServiceOperation serviceOperation = getLastDeleteServiceOperation(execution, service);
+            if (serviceOperation != null) {
+                return serviceOperation;
+            }
+            handleMissingServiceInstance(execution, service);
+            return null;
+        }
+        return getLastOperation(cloudServiceInstance);
+    }
+
+    private ServiceOperation getLastDeleteServiceOperation(ExecutionWrapper execution, CloudServiceExtended service) {
```

Comment:
Isn't this a code duplication?
Please find all duplications in this pull request and try to fix them.